README.md
## Overview

Docker image packaging for oxAuth.

## Versions

- Stable: `gluufederation/oxauth:4.1.3_03`
- Unstable: `gluufederation/oxauth:4.1.3_dev`

Refer to [Changelog](./CHANGES.md) for details on new features, bug fixes, or older releases.

## Environment Variables

The following environment variables are supported by the container:

- `GLUU_CONFIG_ADAPTER`: The config backend adapter, can be `consul` (default) or `kubernetes`.
- `GLUU_CONFIG_CONSUL_HOST`: hostname or IP of Consul (default to `localhost`).
- `GLUU_CONFIG_CONSUL_PORT`: port of Consul (default to `8500`).
- `GLUU_CONFIG_CONSUL_CONSISTENCY`: Consul consistency mode (choose one of `default`, `consistent`, or `stale`). Default to `stale` mode.
- `GLUU_CONFIG_CONSUL_SCHEME`: supported Consul scheme (`http` or `https`).
- `GLUU_CONFIG_CONSUL_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_CONFIG_CONSUL_CACERT_FILE`: path to Consul CA cert file (default to `/etc/certs/consul_ca.crt`). This file will be used if it exists and `GLUU_CONFIG_CONSUL_VERIFY` set to `true`.
- `GLUU_CONFIG_CONSUL_CERT_FILE`: path to Consul cert file (default to `/etc/certs/consul_client.crt`).
- `GLUU_CONFIG_CONSUL_KEY_FILE`: path to Consul key file (default to `/etc/certs/consul_client.key`).
- `GLUU_CONFIG_CONSUL_TOKEN_FILE`: path to file contains ACL token (default to `/etc/certs/consul_token`).
- `GLUU_CONFIG_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_CONFIG_KUBERNETES_CONFIGMAP`: Kubernetes configmaps name (default to `gluu`).
- `GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_SECRET_ADAPTER`: The secrets adapter, can be `vault` or `kubernetes`.
- `GLUU_SECRET_VAULT_SCHEME`: supported Vault scheme (`http` or `https`).
- `GLUU_SECRET_VAULT_HOST`: hostname or IP of Vault (default to `localhost`).
- `GLUU_SECRET_VAULT_PORT`: port of Vault (default to `8200`).
- `GLUU_SECRET_VAULT_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_SECRET_VAULT_ROLE_ID_FILE`: path to file contains Vault AppRole role ID (default to `/etc/certs/vault_role_id`).
- `GLUU_SECRET_VAULT_SECRET_ID_FILE`: path to file contains Vault AppRole secret ID (default to `/etc/certs/vault_secret_id`).
- `GLUU_SECRET_VAULT_CERT_FILE`: path to Vault cert file (default to `/etc/certs/vault_client.crt`).
- `GLUU_SECRET_VAULT_KEY_FILE`: path to Vault key file (default to `/etc/certs/vault_client.key`).
- `GLUU_SECRET_VAULT_CACERT_FILE`: path to Vault CA cert file (default to `/etc/certs/vault_ca.crt`). This file will be used if it exists and `GLUU_SECRET_VAULT_VERIFY` set to `true`.
- `GLUU_SECRET_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_SECRET_KUBERNETES_CONFIGMAP`: Kubernetes secrets name (default to `gluu`).
- `GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_WAIT_MAX_TIME`: How long the startup "health checks" should run (default to `300` seconds).
- `GLUU_WAIT_SLEEP_DURATION`: Delay between startup "health checks" (default to `10` seconds).
- `GLUU_MAX_RAM_PERCENTAGE`: Used in conjunction with Docker memory limitations (`docker run -m <mem>`) to identify the percentage of the maximum amount of heap memory.
- `GLUU_DEBUG_PORT`: port of remote debugging (if omitted, remote debugging will be disabled).
- `GLUU_PERSISTENCE_TYPE`: Persistence backend being used (one of `ldap`, `couchbase`, or `hybrid`; default to `ldap`).
- `GLUU_PERSISTENCE_LDAP_MAPPING`: Specify data that should be saved in LDAP (one of `default`, `user`, `cache`, `site`, or `token`; default to `default`). Note this environment only takes effect when `GLUU_PERSISTENCE_TYPE` is set to `hybrid`.
- `GLUU_LDAP_URL`: Address and port of LDAP server (default to `localhost:1636`); required if `GLUU_PERSISTENCE_TYPE` is set to `ldap` or `hybrid`.
- `GLUU_COUCHBASE_URL`: Address of Couchbase server (default to `localhost`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_USER`: Username of Couchbase server (default to `admin`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CERT_FILE`: Couchbase root certificate location (default to `/etc/certs/couchbase.crt`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_PASSWORD_FILE`: Path to file contains Couchbase password (default to `/etc/gluu/conf/couchbase_password`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CONN_TIMEOUT`: Connect timeout used when a bucket is opened (default to `10000` milliseconds).
- `GLUU_COUCHBASE_CONN_MAX_WAIT`: Maximum time to wait before retrying connection (default to `20000` milliseconds).
- `GLUU_COUCHBASE_SCAN_CONSISTENCY`: Default scan consistency; one of `not_bounded`, `request_plus`, or `statement_plus` (default to `not_bounded`).
- `GLUU_SYNC_CASA_MANIFESTS`: Whether to sync Casa manifest files to Casa containers (default to `false`).
- `GLUU_CASAWATCHER_INTERVAL`: Interval between Casa files synchronization (default to `10`).
- `GLUU_CONTAINER_METADATA`: The name of scheduler to pull container metadata (one of `docker` or `kubernetes`; default to `docker`).
- `GLUU_SYNC_JKS_ENABLED`: Whether to enable JKS sync (default to `false`). Enable this flag if using `gluufederation/key-rotation:4.1.1_02` and below (for key-regeneration).
- `GLUU_SYNC_JKS_INTERVAL`: Interval of JKS sync in seconds (if needed); obsolete.


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 4.1.3_03

Released on September 11th, 2020.

* Upgraded `oxauth-server` build date `2020-09-11 09:36`.

## Version 4.1.3_02

Released on August 26th, 2020.

* Re-enable periodical keystore sync (ref: https://github.com/GluuFederation/docker-oxauth/issues/26).

## Version 4.1.3_01

Released on August 13th, 2020.

* Upgraded `oxauth-server`.
* Update methods to store oxExternalUid as multivalued by default
* Add index for multivalued oxExternalUid

## Version 4.1.2_01

Released on July 11th, 2020.

* Upgraded `oxauth-server`.
* Add support to use any client authentication for revoke token endpoint

## Version 4.1.1_09

Released on June 5th, 2020.

* Upgraded `oxauth-server`.
* Changed logging level for RClone wrapper script.

## Version 4.1.1_08

Released on May 21st, 2020.

* Fixed Curl package installation.

## Version 4.1.1_07

Released on May 20th, 2020.

* Added Curl package.
* Added support for Jackrabbit (Java Content Repository/JCR).
* Added RClone to interact with Jackrabbit via webdav protocol.

## Version 4.1.1_06

Released on April 29th, 2020.

* Added Jython site-packages.

## Version 4.1.1_05

Released on April 20th, 2020.

* Upgraded `oxauth-server`.

## Version 4.1.1_04

Released on April 17th, 2020.

* Upgraded `oxauth-server`.

* Upgrade Jython package.

* Conform to changes in key rotation logic . Fixes [Key Rotation issue #7](https://github.com/GluuFederation/docker-key-rotation/issues/7) in commit https://github.com/GluuFederation/docker-oxauth/commit/16cd18d34304d0ce91f0e4598fbd6a91d8cf2d0f

  1. oxAuth starts off using old/expired JWKS + JKS

  1. Key-rotation re-generates new JWKS + JKS

  1. Key-rotation creates a backup of old JKS and JWKS in oxAuth at /etc/certs/oxauth-keys.jks and /etc/certs/oxauth-keys.json. Note: At this point oxAuth is still using cached (old/expired) JWKS + JKS.

  1. Key-rotation pushes new JKS .Note: At this point oxAuth is still using cached (old/expired) JWKS + JKS.

  1. Key-rotation saves the JWKS in persistence:

    - if the process fails key rotation restores the JSK from backup in oxAuth

    - if the process succeeds, oxAuth reloads itself hence loading new JWKS + JKS pair


## Version 4.1.1_03

Released on April 2nd, 2020.

* Upgraded `oxauth-server`.
* Added `casawatcher` script.

## Version 4.1.1_02

Released on March 25th, 2020.

* Upgraded `oxauth-server` v4.1.1 build at 2020-03-25 to address issue with key regeneration interval (https://github.com/GluuFederation/oxAuth/issues/1299).

## Version 4.1.1_01

Released on March 24th, 2020.

* Conformed to Gluu Server v4.1.1.

## Version 4.1.0_01

Released on March 5th, 2020.

* Conformed to Gluu Server v4.1.

## Version 4.0.1_08

Released on March 5th, 2020.

* Upgraded `oxauth-server` v4.0.sp1.
* Added ENV for customizing Couchbase connection and scan consistency.

## Version 4.0.1_07

Released on January 22nd, 2020.

* Upgraded `oxauth-server` v4.0.Final.patch1 build at 2020-01-20.

## Version 4.0.1_06

Released on December 25th, 2019.

* Pulled SCIM RP and RS client keystores and JWKS files upon container deployment. Reference: https://github.com/GluuFederation/enterprise-edition/issues/22.

## Version 4.0.1_05

Released on December 1st, 2019.

* Upgraded `oxauth-server` v4.0.1.Final build at 2019-11-30.

## Version 4.0.1_04

Released on November 19th, 2019.

* Upgraded to oxAuth v4.0.1.Final build at 2019-11-18.

## Version 4.0.1_03

Released on November 15th, 2019.

* Upgraded to oxAuth v4.0.1.Final build at 2019-11-14.

## Version 4.0.1_02

Released on November 14th, 2019.

* Upgraded `pygluu-containerlib` to show connection issue with Couchbase explicitly.
* Upgraded to oxAuth v4.0.1.Final build at 2019-11-13.

## Version 4.0.1_01

Released on November 1st, 2019.

* Upgraded to Gluu Server 4.0.1.

## Version 4.0.0_01

Released on October 22nd, 2019.

* Upgraded to Gluu Server 4.0.
* Fixed bugs where `twilio.jar` is deleted when there is volume mounted to `/opt/gluu/jetty/oxauth/custom/libs` directory.

## Version 3.1.6_02

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.6_01

Released on April 29th, 2019.

* Upgraded to Gluu Server 3.1.6.

## Version 3.1.5_04

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.5_03

* Added `GLUU_SYNC_JKS_INTERVAL` env var.

Released on April 23rd, 2019.

* Added license info on container startup.

## Version 3.1.5_02

Released on April 9th, 2019.

* Added license info on container startup.
* Disabled `sendServerVersion` config of Jetty server.

## Version 3.1.5_01

Released on March 23rd, 2019.

* Upgraded to Gluu Server 3.1.5.

## Version 3.1.4_03

Released on February 16th, 2019.

* Added missing `idp-signing.crt` and `passport-rp.jks` files.

## Version 3.1.4_02

Released on January 16th, 2019.

* Added `http-forwarded` module to Jetty.

## Version 3.1.4_01

Released on November 12th, 2018.

* Upgraded to Gluu Server 3.1.4.

## Version 3.1.3_05

Released on September 18th, 2018.

* Changed base image to use Alpine 3.8.1.

## Version 3.1.3_04

Released on September 12th, 2018.

* Added feature to connect to secure Consul (HTTPS).

## Version 3.1.3_03

Released on August 31st, 2018.

* Added Tini to handle signal forwarding and reaping zombie processes.

## Version 3.1.3_02

Released on July 20th, 2018.

* Added wrapper to manage config via Consul KV or Kubernetes configmap.

## Version 3.1.3_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.3.

## Version 3.1.2_01

Released on June 6th, 2018.

* Upgraded to Gluu Server 3.1.2.

## Version 3.1.1_rev1.0.0-beta3

Released on March 14th, 2018.

* Added feature to enable/disable remote debugging of JVM.

## Version 3.1.1_rev1.0.0-beta2

Released on October 11th, 2017.

* Use latest oxauth-server build.


scripts/wait_for.py
import argparse
import base64
import json
import logging
import os
import random
import sys
import time

import ldap3
import pyDes

from cbm import CBM
from gluulib import get_manager

logger = logging.getLogger("wait_for")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(name)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)


def decode_password(manager, password_key, salt_key):
    encoded_password = manager.secret.get(password_key)
    encoded_salt = manager.secret.get(salt_key)

    cipher = pyDes.triple_des(
        b"{}".format(encoded_salt),
        pyDes.ECB,
        padmode=pyDes.PAD_PKCS5
    )
    encrypted_text = b"{}".format(base64.b64decode(encoded_password))
    return cipher.decrypt(encrypted_text)


def wait_for_config(manager, max_wait_time, sleep_duration):
    for i in range(0, max_wait_time, sleep_duration):
        try:
            reason = "config 'hostname' is not available"
            if manager.config.get("hostname"):
                logger.info("Config backend is ready.")
                return
        except Exception as exc:
            reason = exc

        logger.warn("Config backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Config backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for_secret(manager, max_wait_time, sleep_duration):
    for i in range(0, max_wait_time, sleep_duration):
        try:
            reason = "secret 'ssl_cert' is not available"
            if manager.secret.get("ssl_cert"):
                logger.info("Secret backend is ready.")
                return
        except Exception as exc:
            reason = exc

        logger.warn("Secret backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Secret backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for_ldap(manager, max_wait_time, sleep_duration):
    GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "localhost:1636")
    GLUU_PERSISTENCE_LDAP_MAPPING = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")

    ldap_bind_dn = manager.config.get("ldap_binddn")
    ldap_password = decode_password(manager, "encoded_ox_ldap_pw", "encoded_salt")

    ldap_host = GLUU_LDAP_URL.split(":")[0]
    ldap_port = int(GLUU_LDAP_URL.split(":")[1])

    ldap_server = ldap3.Server(
        ldap_host,
        ldap_port,
        use_ssl=True
    )

    # check the entries few times, to ensure OpenDJ is running after importing
    # initial data; this may not required for OpenLDAP
    successive_entries_check = 0

    search_base_mapping = {
        "default": "o=gluu",
        "user": "o=gluu",
        "site": "o=site",
        "cache": "o=gluu",
        "statistic": "o=metric",
    }
    search_base = search_base_mapping[GLUU_PERSISTENCE_LDAP_MAPPING]

    for i in range(0, max_wait_time, sleep_duration):
        try:
            with ldap3.Connection(
                    ldap_server,
                    ldap_bind_dn,
                    ldap_password) as ldap_connection:

                ldap_connection.search(
                    search_base=search_base,
                    search_filter="(objectClass=*)",
                    search_scope=ldap3.SUBTREE,
                    attributes=['objectClass'],
                    size_limit=1,
                )

                if ldap_connection.entries:
                    successive_entries_check += 1

                if successive_entries_check >= 3:
                    logger.info("LDAP is ready")
                    return
                reason = "LDAP is not initialized yet"
        except Exception as exc:
            reason = exc

        logger.warn("LDAP backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("LDAP not ready, after " + str(max_wait_time) + " seconds.")
    sys.exit(1)


def check_couchbase_document(cbm):
    persistence_type = os.environ.get("GLUU_PERSISTENCE_TYPE", "ldap")
    ldap_mapping = os.environ.get("GLUU_PERSISTENCE_LDAP_MAPPING", "default")
    checked = True
    error = ""
    bucket = "gluu"

    if persistence_type == "hybrid":
        req = cbm.get_buckets()
        if not req.ok:
            checked = False
            error = json.loads(req.text)["errors"][0]["msg"]
            return checked, error

        bucket = random.choice([
            _bucket["name"] for _bucket in req.json()
            if _bucket["name"] != ldap_mapping
        ])

    query = "SELECT COUNT(*) FROM `{}`".format(bucket)
    req = cbm.exec_query(query)
    if not req.ok:
        checked = False
        error = json.loads(req.text)["errors"][0]["msg"]
    return checked, error


def wait_for_couchbase(manager, max_wait_time, sleep_duration):
    hostname = os.environ.get("GLUU_COUCHBASE_URL", "localhost")
    user = manager.config.get("couchbase_server_user")
    cbm = CBM(hostname, user, decode_password(
        manager, "encoded_couchbase_server_pw", "encoded_salt",
    ))

    for i in range(0, max_wait_time, sleep_duration):
        try:
            checked, error = check_couchbase_document(cbm)
            if checked:
                logger.info("Couchbase is ready")
                return
            reason = error
        except Exception as exc:
            reason = exc

        logger.warn("Couchbase backend is not ready; reason={}; "
                    "retrying in {} seconds.".format(reason, sleep_duration))
        time.sleep(sleep_duration)

    logger.error("Couchbase backend is not ready after {} seconds.".format(max_wait_time))
    sys.exit(1)


def wait_for(manager, deps=None):
    deps = deps or []

    try:
        max_wait_time = int(os.environ.get("GLUU_WAIT_MAX_TIME", 300))
    except ValueError:
        max_wait_time = 300

    try:
        sleep_duration = int(os.environ.get("GLUU_WAIT_SLEEP_DURATION", 5))
    except ValueError:
        sleep_duration = 5

    if "config" in deps:
        wait_for_config(manager, max_wait_time, sleep_duration)

    if "secret" in deps:
        wait_for_secret(manager, max_wait_time, sleep_duration)

    if "ldap" in deps:
        wait_for_ldap(manager, max_wait_time, sleep_duration)

    if "couchbase" in deps:
        wait_for_couchbase(manager, max_wait_time, sleep_duration)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--deps", help="Comma-separated dependencies to wait for.")
    args = parser.parse_args()

    deps = set(filter(
        None,
        [dep.strip() for dep in args.deps.split(",") if dep]
    ))

    manager = get_manager()
    wait_for(manager, deps)
